### ISSUE TITLE
Avoiding edit wars

### ISSUE THREAD

**Issue by Dugy** [Dugy/Legend_of_the_Invincibles#223]
I have noticed that my style of writing lua differs from @edwardspec 's style, which almost caused me to start an edit war at `stats.cfg`.

I am trying to suggest a unified code style recommended for this add-on, because we have some mess here already.

# Function names:

Should be `loti.unit.something()`, `loti.item.something()`, so they should be in the `loti` namespace and then grouped if they pertain more to units or to items. I was doing this differently, but now I noticed this is a better approach.

# WML tags

WML tags can be accessed using `wml.tag.something` or using `T.something` after defining T as a WML tag metatable. Using `T.something` is better because it makes the code shorter, which makes a big change in sections where it's used very often (such as GUI layout definitions).

# GUI definitions
A GUI definition is a load of repetitive WML code written in lua. Its structure of nesting tags resembles HTML a lot. It often nests too deep, creating hadouken code.

It is possible to set a few nested blocks to separate variables and then include them on one line, making the code look better. It doesn't hide errors if each of these variables can be put into a column (that is, either a grid or a widget, not rows or columns). However, while trying to decode a recent version of `stats.cfg`, I found that it's not exactly the best approach. The layout absolutely cannot fit into a Full HD window (not probably not into a 4K either), because horizontal space is wasted. Also, it's split into many parts that have to be looked up and finding which part is where on the GUI needs to use the Search function. Also, one line can unexpectedly represent a large block.

Therefore, please write tag attributes on the same line to save vertical space and split parts of the tree into separate variables or generating functions only if it helps avoid boilerplate code.

# Questions?

Of course, if you have other reasons to write the code this way, write them here, I might not have had the problems with the other approach only because I have not been reading something written that way.

**Comment by edwardspec** [comment]
> so they should be in the loti namespace and then grouped if they pertain more to units or to items

Agreed.
There may be other categories (e.g. I grouped "gem counts" into loti.gem, so that they can be named get_counts, set_counts -  crafting dialog functions are also in loti.gem), if it makes function names more short or easy to remember.

> Using T.something is better

Agreed.
It's just an alias made with `T = wml.tag`, so a shorter version is preferable.
(`helper.set_wml_tag_metatable` method is deprecated, but it doesn't matter)

> It is possible to set a few nested blocks to separate variables and then include them on one line, making the code look better.

Can you provide an example? (code written this way)

**Comment by Dugy** [comment]
> There may be other categories (e.g. I grouped "gem counts" into loti.gem, so that they can be named get_counts, set_counts - crafting dialog functions are also in loti.gem), if it makes function names more short or easy to remember.

Okay, as long as there are not categories for only one or two functions.

> helper.set_wml_tag_metatable method is deprecated
Aha. I see now.

> Can you provide an example? (code written this way)

A piece of your code in `crafting.cfg`. I have skipped one part to make it shorter.
```lua
	local yesno_buttons =  wml.tag.grid {
		wml.tag.row {
			wml.tag.column {
				wml.tag.button {
					id = "ok",
					label = _"Craft"
				}
			},
			wml.tag.spacer {},
			wml.tag.column {
				wml.tag.button {
					id = "cancel",
					label = _"Back"
				}
			},
			wml.tag.spacer {},
			wml.tag.column {
				wml.tag.button {
					id = "transmute",
					label = _"Transmute gems"
				}
			}
		}
	}

	return {
		wml.tag.tooltip { id = "tooltip_large" },
		wml.tag.helptip { id = "tooltip_large" },
		maximum_width = 1000,
		maximum_height = 800,
		wml.tag.grid {
			wml.tag.row {
				wml.tag.column {
					border = "bottom",
					border_size = 10,
					wml.tag.label {
						definition = "title",
						label = _"Crafting"
					}
				}
			},
			wml.tag.row {
				wml.tag.column {
					border = "top",
					border_size = 10,
					horizontal_grow = true,
					yesno_buttons
				}
			}
		}
	}
```

Which could be written as:
```lua
	return {
		wml.tag.tooltip { id = "tooltip_large" },
		wml.tag.helptip { id = "tooltip_large" },
		maximum_width = 1000,
		maximum_height = 800,
		wml.tag.grid {
			wml.tag.row {
				wml.tag.column {
					border = "bottom",
					border_size = 10,
					wml.tag.label {
						definition = "title",
						label = _"Crafting"
					}
				}
			},
			wml.tag.row {
				wml.tag.column {
					border = "top",
					border_size = 10,
					horizontal_grow = true,
					wml.tag.grid {
						wml.tag.row {
							wml.tag.column {
								wml.tag.button {
									id = "ok",
									label = _"Craft"
								}
							},
							wml.tag.spacer {},
							wml.tag.column {
								wml.tag.button {
									id = "cancel",
									label = _"Back"
								}
							},
							wml.tag.spacer {},
							wml.tag.column {
								wml.tag.button {
									id = "transmute",
									label = _"Transmute gems"
								}
							}
						}
					}
				}
			}
		}
	}
```
It makes the tree less scattered across the file. But it's still impossible to see much of the code on one screen.

It could be further changed to:
``` lua
	return {
		wml.tag.tooltip { id = "tooltip_large" }, wml.tag.helptip { id = "tooltip_large" }, maximum_width = 1000, maximum_height = 800,
		wml.tag.grid {
			wml.tag.row {
				wml.tag.column { border = "bottom", border_size = 10,
					wml.tag.label { definition = "title", label = _"Crafting" }
				}
			},
			wml.tag.row {
				wml.tag.column { border = "top", border_size = 10, horizontal_grow = true,
					wml.tag.grid {
						wml.tag.row {
							wml.tag.column {
								wml.tag.button { id = "ok", label = _"Craft" }
							},
							wml.tag.spacer {},
							wml.tag.column {
								wml.tag.button { id = "cancel", label = _"Back" }
							},
							wml.tag.spacer {},
							wml.tag.column {
								wml.tag.button { id = "transmute", label = _"Transmute gems" }
							}
						}
					}
				}
			}
		}
	}
```
It keeps the tree structure readable and allows much more stuff to fit into a screen (especially into a text editor with roughly HD size). The tree structure remains visible. It's similarly readable than a piece of code in a C-like programming language, where the curly brackets are like blocks of code and the attributes are like function arguments.

I was probably going into the opposite extreme, compressing too much stuff on individual lines.

**Comment by edwardspec** [comment]
The compact version does indeed help on large screens, but the price is readability of diffs.
Examples are below.

The thing is, there aren't many situations when we need to look at the code of entire dialog (when seeing it all on 1 screen would be helpful) rather than tweak one specific part of it. Only when originally designing a new dialog. All further changes will benefit from cleaner diffs more.

1) Variables like yesno_buttons or basetype_listbox make it easy to rearrange the widgets.

Suppose you want to move some listbox to a different column. If it's not a variable, then the diff would show a huge amount of code deleted from one place and inserted into another.
The worst part is, it may be difficult to see whether the added text is exactly the same as deleted text, or if there were some subtle changes in it (e.g. refactoring in the same commit as moving the column - possibly introducing a new typo/bug).

If they are variables, you see a commit like this: https://github.com/Dugy/Legend_of_the_Invincibles/commit/738ccba495109946e94797f90e95defb238688fe - you can immediately see what happened to `type_listbox` (where it was deleted from, where it is now, and that its contents haven't changed).

2) I very much recommend against writing several properties in one line.

A very common type of change (in existing code) is to tweak one of the properties (e.g. increase border), or add an additional property.
In this situation, "properties on one line" code style would show this:

```diff
- wml.tag.tooltip { id = "tooltip_large" }, wml.tag.helptip { id = "tooltip_large" }, maximum_width = 1000, maximum_height = 800,
+ wml.tag.tooltip { id = "tooltip_large" }, wml.tag.helptip { id = "tooltip_large" }, maximum_width = 1100, maximum_height = 800, definition = "wml_message",
```

... and "each property on a line" would show this:
```diff
- maximum_width = 1000,
+ maximum_width = 1100,
+ definition = "wml_message",
```

With tools like wdiff or gitk, changed parts of the long line can be determined (in green/red for added/deleted words), but in many situations a full-line diff will be shown.
Diffs like https://github.com/Dugy/Legend_of_the_Invincibles/commit/2b954ad0929682b496e71c37054658dc8353d3af may be unavoidable for text, but we can avoid them for code (properties).

Another thing is that writing properties in separate lines avoids merge conflicts.

Suppose you make a commit that changes maximum_width of dialog, dalong76 makes a commit that changes label of the button, and my commit adds horizontal_alignment parameter somewhere. If properties are written on their lines, then Git is very likely to resolve this automatically, unless the changes affect same/neighboring lines.
However, in more compact code style, this may cause up to two merge conflicts (having properties in the same line increases the possibility that changes lines will be too close for automatic resolving) that would need to be resolved by humans.

And again, there are few situations when you need to look at a large portion of the dialog rather at a specific part of it (which will fit on the screen even if properties are on their own lines).

I think that there is no problem writing things like `wml.tag.button { id = "ok", label = _"Craft" }` (which is unlikely to be changed) on one line,
but writing things with more parameters (especially {horizontal,vertical}_{alignment,grow}, border{,_size}, etc. - that are likely to be changed) in one line is just asking for unnoticed typos.

Writing properties on different lines could have been an issue if one needed to scroll a lot to modify some part of the code, but it's not the case. Even on my 1366x768 notebook, I can see `listbox_template` and `listbox` of Transmuting dialog fitting in one screen. `basetype_listbox` fits in one screen. Top-level dialog (first parameter to `wesnoth.show_dialog`) fits in one screen.

> HTML, XML or JSON, where attributes are not written on separate lines.

Not exactly true.
When using JSON for anything human-readable (configs, etc.), attributes are always written on separate lines (for the same reasons I mentioned above - easily tracking changes, avoiding merge conflicts). Keeping a tonn of properties in HTML is an antipattern (having HTML with many properties in Git repository would result in ugly diffs), and now that most of them got moved into CSS, they are also written one per line there.

**Comment by Dugy** [comment]
Shortly after writing that above, I realised that JSON is usually not written with more attributes on one line, but I was not anywhere near my laptop and then I forgot about it. The mistake is particularly strange because shortly before writing it, I was writing a draft of a JSON-based communication protocol. JSON might not be a good comparison overall because it doesn't support comments and thus no explanations can't be put anywhere.

The readability of the code itself is more important than the readability of diffs. Typical workflow is doing your work and make a pull rebase occasionally. Reading diffs is important only if something goes wrong. And actually its effect on diff readability is low, the difference in one property is perfectly readable even if there are many properties on each line (as in your example).

The properties themselves are of little relevance over the whole structure. We don't write one function argument per line even in languages where it can be done.

It's unlikely there will be some merge conflicts because of this, because two or more people would have to be editing the same part of the same window, which would have to be manually merged even without merge conflicts.

You have a very small display, especially with little vertical space. I see that it's why you put some parts like the yesno_buttons into separate parts to solve the problems with excessive vertical space consumption. It means that something of the whole is visible in one screen, but at the cost of scattering the code over the file. The code could fit into one screen if it was more compact, but this way I had to jump from one place to another and scroll up and down all the time to understand what you wrote. If you have written it, it's likely that you remember what is where, but everyone else has difficulties. Again, it would not matter if there was a function to create those listboxes or yes/no buttons used in multiple locations (which might be needed, actually), because it would be remembered after being read at one location and its only alternative is boilerplate code.

In short, you are slightly improving the readability of diffs and marginally reducing the probability of merge conflicts at the cost of large code readability problems. This reminds me of cargo cult practices.

So I agree to place properties that are likely to be changed often, such as maximum_height and maximum_width, on a separate line (in groups of arguments likely to be changed together). Otherwise, please don't sacrifice readability of code itself for minor improvements of readability of diffs.

**Comment by edwardspec** [comment]
Very well. Feel free to rearrange the code of existing dialogs as convenient for you.
Let's delay this on the `inventory` branch until it is completed and merged into master. (Inventory dialog is a very large feature - more than 50 commits - and I'd rather not reformat its code drastically, as it would delay the development)

**Comment by Dugy** [comment]
Fine, I will not edit it and I will leave it alone until it's in master. I will change crafting when I'll need to edit it in some way.

**Comment by edwardspec** [comment]
>  when I'll need to edit it in some way.

Note that this should be in a separate commit from functional changes.
Not like "this 1 commit rearranges 50 lines and also slightly modifies 3 of them, but we would have no idea where exactly they were modified by looking at a diff".

**Comment by Dugy** [comment]
Okay.

**Comment by edwardspec** [comment]
Another code style difference I wanted to discuss is
```lua
for i = 1,#array do
	many_lines_that_work_with(array[i])
end
```
vs.
```lua
for i, elem in ipairs(array) do
	many_lines_that_work_with(elem)
end
```

I've been replacing 1 with 2 in places where 1 was causing code repetitions (potentially typo-prone) inside the loop: multiple `[i]` in the code that doesn't really depend on the index (doesn't change the element of array, etc.).

Example from the real code:
```lua
if specials[i][1] == "damage" and string.match(specials[i][2].id, "backstab") then
	local quality = specials[i][2].multiply
	if quality then
		if specials[i][2].apply_to ~= "both" then
			quality = quality * 0.75 -- Charging backstab is worse
		end
		if quality > best_backstab_mult then
			best_backstab_mult = quality
			best_backstab = specials[i]
```
With array[i] being a WML table (where we often check `[1]` and `[2]`), there is even more possibility to accidentally write `specials[1]` instead of `specials[i]`, etc. I actually caught one such typo in the past.
